File: apps/server/engine/event_bus.py
"""
Event Bus for AITuberFlow

Manages event routing between nodes with support for event filtering.
"""

from typing import Dict, List, Callable, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime
import asyncio
import logging
import re

logger = logging.getLogger(__name__)


@dataclass
class Event:
    """Represents an event in the system."""
    type: str
    payload: Dict[str, Any]
    source_node_id: Optional[str] = None
    timestamp: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class EventFilter:
    """Defines a filter for events."""
    event: str  # Event type pattern (supports wildcards like "message.*")
    condition: Optional[str] = None  # JavaScript-like condition expression

    def matches(self, event: Event) -> bool:
        """Check if an event matches this filter."""
        # Check event type pattern
        if not self._match_pattern(event.type):
            return False

        # Check condition if specified
        if self.condition:
            return self._evaluate_condition(event)

        return True

    def _match_pattern(self, event_type: str) -> bool:
        """Match event type against pattern (supports wildcards)."""
        pattern = self.event
        # Convert wildcard pattern to regex
        if '*' in pattern:
            regex = pattern.replace('.', r'\.').replace('*', '.*')
            return bool(re.match(f'^{regex}$', event_type))
        return pattern == event_type

    def _evaluate_condition(self, event: Event) -> bool:
        """Evaluate a condition expression against event payload."""
        if not self.condition:
            return True

        try:
            # Create evaluation context
            context = {
                'event': event.payload,
                'type': event.type,
                'source': event.source_node_id,
            }

            # Simple expression evaluation (safe subset)
            # Supports: event.field, comparisons, and/or
            condition = self.condition

            # Replace event.field with actual values
            # Note: bool must be checked before int (bool is subclass of int)
            for key, value in event.payload.items():
                if isinstance(value, bool):
                    condition = condition.replace(f'event.{key}', str(value))
                elif isinstance(value, str):
                    condition = condition.replace(f'event.{key}', f'"{value}"')
                elif isinstance(value, (int, float)):
                    condition = condition.replace(f'event.{key}', str(value))

            # Evaluate simple expressions
            # Support: ==, !=, >, <, >=, <=, and, or, not
            condition = condition.replace('&&', ' and ').replace('||', ' or ')
            condition = condition.replace('===', '==').replace('!==', '!=')
            # Support JavaScript-style true/false
            condition = condition.replace('true', 'True').replace('false', 'False')

            # Safe eval with restricted builtins
            result = eval(condition, {"__builtins__": {}}, {"True": True, "False": False})
            return bool(result)

        except Exception as e:
            logger.warning(f"Failed to evaluate condition '{self.condition}': {e}")
            return False


@dataclass
class Subscription:
    """Represents a subscription to events."""
    callback: Callable
    filters: List[EventFilter] = field(default_factory=list)
    node_id: Optional[str] = None


class EventBus:
    """
    Event bus for managing workflow events.
    Supports event filtering and pattern matching.
    """

    def __init__(self):
        self._subscriptions: Dict[str, List[Subscription]] = {}
        self._event_queue: asyncio.Queue = None
        self._running = False
        self._event_history: List[Event] = []
        self._max_history = 100

    async def start(self):
        """Start the event bus."""
        self._event_queue = asyncio.Queue()
        self._running = True
        self._event_history = []
        logger.info("Event bus started")

    async def stop(self):
        """Stop the event bus."""
        self._running = False
        if self._event_queue:
            while not self._event_queue.empty():
                try:
                    self._event_queue.get_nowait()
                except asyncio.QueueEmpty:
                    break
        logger.info("Event bus stopped")

    def subscribe(
        self,
        event_type: str,
        callback: Callable,
        filters: List[EventFilter] = None,
        node_id: str = None
    ) -> str:
        """
        Subscribe to an event type with optional filters.

        Args:
            event_type: Event type pattern (supports wildcards)
            callback: Async callback function
            filters: Optional list of additional filters
            node_id: Optional node ID for tracking

        Returns:
            Subscription ID
        """
        if event_type not in self._subscriptions:
            self._subscriptions[event_type] = []

        subscription = Subscription(
            callback=callback,
            filters=filters or [],
            node_id=node_id
        )
        self._subscriptions[event_type].append(subscription)

        sub_id = f"{event_type}:{len(self._subscriptions[event_type])-1}"
        logger.debug(f"Subscribed to event: {event_type} (node: {node_id})")
        return sub_id

    def unsubscribe(self, event_type: str, callback: Callable = None, node_id: str = None):
        """Unsubscribe from an event type."""
        if event_type not in self._subscriptions:
            return

        if callback:
            self._subscriptions[event_type] = [
                sub for sub in self._subscriptions[event_type]
                if sub.callback != callback
            ]
        elif node_id:
            self._subscriptions[event_type] = [
                sub for sub in self._subscriptions[event_type]
                if sub.node_id != node_id
            ]

    async def emit(self, event: Event) -> int:
        """
        Emit an event to all matching subscribers.

        Returns:
            Number of subscribers notified
        """
        if not self._running:
            logger.warning("Event bus is not running, event dropped")
            return 0

        logger.debug(f"Emitting event: {event.type}")

        # Store in history
        self._event_history.append(event)
        if len(self._event_history) > self._max_history:
            self._event_history = self._event_history[-self._max_history:]

        notified = 0

        # Find matching subscriptions
        for pattern, subscriptions in self._subscriptions.items():
            if not self._pattern_matches(pattern, event.type):
                continue

            for subscription in subscriptions:
                # Check filters
                if subscription.filters:
                    if not all(f.matches(event) for f in subscription.filters):
                        continue

                try:
                    if asyncio.iscoroutinefunction(subscription.callback):
                        await subscription.callback(event)
                    else:
                        subscription.callback(event)
                    notified += 1
                except Exception as e:
                    logger.error(f"Error in event handler for {event.type}: {e}")

        return notified

    def _pattern_matches(self, pattern: str, event_type: str) -> bool:
        """Check if a subscription pattern matches an event type."""
        if pattern == '*':
            return True
        if '*' in pattern:
            regex = pattern.replace('.', r'\.').replace('*', '.*')
            return bool(re.match(f'^{regex}$', event_type))
        return pattern == event_type

    def clear_subscriptions(self, node_id: str = None):
        """Clear subscriptions, optionally for a specific node."""
        if node_id:
            for event_type in self._subscriptions:
                self._subscriptions[event_type] = [
                    sub for sub in self._subscriptions[event_type]
                    if sub.node_id != node_id
                ]
        else:
            self._subscriptions.clear()

    def get_history(self, event_type: str = None, limit: int = 10) -> List[Event]:
        """Get recent event history."""
        if event_type:
            events = [e for e in self._event_history if e.type == event_type]
        else:
            events = self._event_history
        return events[-limit:]


# Global event bus instance
event_bus = EventBus()

File: apps/server-ts/src/engine/event-bus.ts
/**
 * Event Bus for AITuberFlow
 *
 * Manages event routing between nodes with support for
 * wildcard pattern matching and safe condition filtering.
 */

import type { Event } from "@aituber-flow/sdk";
export type { Event } from "@aituber-flow/sdk";

// ─── EventFilter ─────────────────────────────────────────────────

/**
 * Defines a filter for events.
 * Supports wildcard patterns (e.g. "message.*") and
 * safe condition expressions (e.g. "event.amount > 100").
 */
export class EventFilter {
  readonly event: string;
  readonly condition?: string;

  constructor(event: string, condition?: string) {
    this.event = event;
    this.condition = condition;
  }

  matches(event: Event): boolean {
    if (!this._matchPattern(event.type)) return false;
    if (this.condition) return this._checkCondition(event);
    return true;
  }

  /**
   * Match event type against pattern (supports wildcards).
   * "message.*" matches "message.received", "message.sent", etc.
   */
  private _matchPattern(eventType: string): boolean {
    const pattern = this.event;
    if (!pattern.includes("*")) return pattern === eventType;

    const regex = new RegExp(`^${pattern.replace(/\./g, "\\.").replace(/\*/g, ".*")}$`);
    return regex.test(eventType);
  }

  /**
   * Check a condition expression against event payload.
   *
   * Safely parses simple comparison expressions using manual parsing.
   * Supports: event.field references, comparison operators
   * (==, !=, >, <, >=, <=), logical operators (and/&&, or/||).
   */
  private _checkCondition(event: Event): boolean {
    if (!this.condition) return true;

    try {
      // Normalize logical operators
      const normalized = this.condition.replace(/\band\b/g, "&&").replace(/\bor\b/g, "||");

      // Split on logical operators and check each clause
      return this._parseLogical(normalized, event);
    } catch {
      return false;
    }
  }

  /**
   * Parse a logical expression (handles && and ||).
   */
  private _parseLogical(expr: string, event: Event): boolean {
    // Handle || (lowest precedence)
    const orParts = this._splitOnOperator(expr, "||");
    if (orParts.length > 1) {
      return orParts.some((part) => this._parseLogical(part, event));
    }

    // Handle && (higher precedence)
    const andParts = this._splitOnOperator(expr, "&&");
    if (andParts.length > 1) {
      return andParts.every((part) => this._parseLogical(part, event));
    }

    // Single comparison
    return this._parseComparison(expr.trim(), event);
  }

  /**
   * Split expression on an operator, respecting quoted strings.
   */
  private _splitOnOperator(expr: string, op: string): string[] {
    const parts: string[] = [];
    let depth = 0;
    let inString = false;
    let stringChar = "";
    let current = "";

    for (let i = 0; i < expr.length; i++) {
      const char = expr[i];

      if (inString) {
        current += char;
        if (char === stringChar) inString = false;
        continue;
      }

      if (char === '"' || char === "'") {
        inString = true;
        stringChar = char;
        current += char;
        continue;
      }

      if (char === "(") depth++;
      if (char === ")") depth--;

      if (depth === 0 && expr.slice(i, i + op.length) === op) {
        parts.push(current);
        current = "";
        i += op.length - 1;
        continue;
      }

      current += char;
    }

    parts.push(current);
    return parts;
  }

  /**
   * Parse a single comparison expression (e.g. "event.amount > 100").
   */
  private _parseComparison(expr: string, event: Event): boolean {
    // Match: value operator value
    const match = expr.match(/^\s*(.+?)\s*(===|!==|==|!=|>=|<=|>|<)\s*(.+?)\s*$/);
    if (!match) {
      // Try as a single truthy value
      const val = this._resolveValue(expr.trim(), event);
      return Boolean(val);
    }

    const [, leftRaw, operator, rightRaw] = match;
    const left = this._resolveValue(leftRaw, event);
    const right = this._resolveValue(rightRaw, event);

    switch (operator) {
      case "==":
      case "===":
        return left === right;
      case "!=":
      case "!==":
        return left !== right;
      case ">":
        return Number(left) > Number(right);
      case "<":
        return Number(left) < Number(right);
      case ">=":
        return Number(left) >= Number(right);
      case "<=":
        return Number(left) <= Number(right);
      default:
        return false;
    }
  }

  /**
   * Resolve a value reference against the event.
   * Handles: event.field paths, string literals, numbers, booleans.
   */
  private _resolveValue(raw: string, event: Event): string | number | boolean | null {
    const trimmed = raw.trim();

    // String literal
    if (
      (trimmed.startsWith('"') && trimmed.endsWith('"')) ||
      (trimmed.startsWith("'") && trimmed.endsWith("'"))
    ) {
      return trimmed.slice(1, -1);
    }

    // Boolean literals
    if (trimmed === "true" || trimmed === "True") return true;
    if (trimmed === "false" || trimmed === "False") return false;

    // Null
    if (trimmed === "null" || trimmed === "None") return null;

    // Number
    if (/^-?\d+(\.\d+)?$/.test(trimmed)) return Number(trimmed);

    // Event payload property path (e.g. "event.amount" or "event.user.name")
    if (trimmed.startsWith("event.")) {
      const path = trimmed.slice(6); // strip "event."
      return this._resolvePath(event.payload, path);
    }

    // Bare string fallback
    return trimmed;
  }

  /**
   * Safely resolve a dot-separated property path on an object.
   */
  private _resolvePath(
    obj: Record<string, unknown>,
    path: string,
  ): string | number | boolean | null {
    const parts = path.split(".");
    let current: unknown = obj;
    for (const part of parts) {
      if (current == null || typeof current !== "object") return null;
      current = (current as Record<string, unknown>)[part];
    }
    if (current == null) return null;
    if (
      typeof current === "string" ||
      typeof current === "number" ||
      typeof current === "boolean"
    ) {
      return current;
    }
    return null;
  }
}

// ─── Subscription ────────────────────────────────────────────────

export interface Subscription {
  callback: (event: Event) => void | Promise<void>;
  filters: EventFilter[];
  nodeId?: string;
}

// ─── EventBus ────────────────────────────────────────────────────

/**
 * Event bus for managing workflow events.
 * Supports event filtering and wildcard pattern matching.
 */
export class EventBus {
  private _subscriptions: Map<string, Subscription[]> = new Map();
  private _running = false;
  private _eventHistory: Event[] = [];
  private _maxHistory = 100;

  async start(): Promise<void> {
    this._running = true;
    this._eventHistory = [];
  }

  async stop(): Promise<void> {
    this._running = false;
  }

  /**
   * Subscribe to an event type with optional filters.
   *
   * @param eventType - Event type pattern (supports wildcards)
   * @param callback  - Async or sync callback
   * @param filters   - Optional additional filters
   * @param nodeId    - Optional node ID for tracking
   * @returns Subscription ID
   */
  subscribe(
    eventType: string,
    callback: (event: Event) => void | Promise<void>,
    filters?: EventFilter[],
    nodeId?: string,
  ): string {
    if (!this._subscriptions.has(eventType)) {
      this._subscriptions.set(eventType, []);
    }

    const subs = this._subscriptions.get(eventType);
    if (!subs) return "";
    subs.push({ callback, filters: filters ?? [], nodeId });

    return `${eventType}:${subs.length - 1}`;
  }

  /**
   * Unsubscribe from an event type by callback or nodeId.
   */
  unsubscribe(
    eventType: string,
    callback?: (event: Event) => void | Promise<void>,
    nodeId?: string,
  ): void {
    const subs = this._subscriptions.get(eventType);
    if (!subs) return;

    if (callback) {
      this._subscriptions.set(
        eventType,
        subs.filter((s) => s.callback !== callback),
      );
    } else if (nodeId) {
      this._subscriptions.set(
        eventType,
        subs.filter((s) => s.nodeId !== nodeId),
      );
    }
  }

  /**
   * Emit an event to all matching subscribers.
   *
   * @returns Number of subscribers notified
   */
  async emit(event: Event): Promise<number> {
    if (!this._running) return 0;

    // Store in history
    this._eventHistory.push(event);
    if (this._eventHistory.length > this._maxHistory) {
      this._eventHistory = this._eventHistory.slice(-this._maxHistory);
    }

    let notified = 0;

    for (const [pattern, subscriptions] of this._subscriptions) {
      if (!this._patternMatches(pattern, event.type)) continue;

      for (const sub of subscriptions) {
        if (sub.filters.length > 0) {
          if (!sub.filters.every((f) => f.matches(event))) continue;
        }

        try {
          await sub.callback(event);
          notified++;
        } catch (err) {
          console.error(`Error in event handler for ${event.type}:`, err);
        }
      }
    }

    return notified;
  }

  /**
   * Check if a subscription pattern matches an event type.
   */
  private _patternMatches(pattern: string, eventType: string): boolean {
    if (pattern === "*") return true;
    if (!pattern.includes("*")) return pattern === eventType;

    const regex = new RegExp(`^${pattern.replace(/\./g, "\\.").replace(/\*/g, ".*")}$`);
    return regex.test(eventType);
  }

  /**
   * Clear subscriptions, optionally for a specific node.
   */
  clearSubscriptions(nodeId?: string): void {
    if (nodeId) {
      for (const [eventType, subs] of this._subscriptions) {
        this._subscriptions.set(
          eventType,
          subs.filter((s) => s.nodeId !== nodeId),
        );
      }
    } else {
      this._subscriptions.clear();
    }
  }

  /**
   * Get recent event history, optionally filtered by type.
   */
  getHistory(eventType?: string, limit = 10): Event[] {
    const events = eventType
      ? this._eventHistory.filter((e) => e.type === eventType)
      : this._eventHistory;
    return events.slice(-limit);
  }
}

/** Global event bus instance */
export const eventBus = new EventBus();
